### PR TITLE
Second prev info in rao result

### PR DESCRIPTION
**Pull request by MartinBelthle**
**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*

No

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*

Feature

**What is the current behavior?** *(You can also link to an open issue here)*

You can only know if a rao went through 2nd preventive by reading the logs

**What is the new behavior (if this is a feature change)?**

You can access this information in the raoresult object

**Does this PR introduce a breaking change?** *(What changes might users need to make in their application due to this PR?)*

No

**Other information**:

(if any of the questions/checkboxes don't apply, please delete them entirely)

### CODE REVIEW

**Review by phiedw — CHANGES_REQUESTED** [review]
After discussing with Godelaine, we have come to the conclusion that it would be even better to have an enum to know whether we finished after the first preventive, run a second preventive and kept the 2P result, or run a second preventive and kept the 1P result

**Comment by pet-mit** [comment]
> After discussing with Godelaine, we have come to the conclusion that it would be even better to have an enum to know whether we finished after the first preventive, run a second preventive and kept the 2P result, or run a second preventive and kept the 1P result

Or ran a 1P/2P but fell back to the initial situation

File: ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/result/impl/PreventiveAndCurativesRaoResultImpl.java [hunk 1]
```diff
@@ -438,4 +440,26 @@ private State getStateOptimizedBefore(State state) {
...
+                !(optimizationStepsExecuted.equals(OptimizationStepsExecuted.FIRST_PREVENTIVE_FELLBACK) ||
+                        optimizationStepsExecuted.equals(OptimizationStepsExecuted.SECOND_PREVENTIVE_IMPROVED_FIRST) ||
+                        optimizationStepsExecuted.equals(OptimizationStepsExecuted.SECOND_PREVENTIVE_FELLBACK_TO_FIRST_PREVENTIVE_SITUATION)) ||
+                (this.optimizationStepsExecuted.equals(OptimizationStepsExecuted.SECOND_PREVENTIVE_IMPROVED_FIRST) &&
+                        !optimizationStepsExecuted.equals(OptimizationStepsExecuted.SECOND_PREVENTIVE_FELLBACK_TO_INITIAL_SITUATION)) ||
+                (this.optimizationStepsExecuted.equals(OptimizationStepsExecuted.SECOND_PREVENTIVE_FELLBACK_TO_FIRST_PREVENTIVE_SITUATION) &&
+                        !optimizationStepsExecuted.equals(OptimizationStepsExecuted.SECOND_PREVENTIVE_FELLBACK_TO_INITIAL_SITUATION)) ||
+                this.optimizationStepsExecuted.equals(OptimizationStepsExecuted.FIRST_PREVENTIVE_FELLBACK) ||
+                this.optimizationStepsExecuted.equals(OptimizationStepsExecuted.SECOND_PREVENTIVE_FELLBACK_TO_INITIAL_SITUATION)) {
```
Comment (phiedw) [inline ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/result/impl/PreventiveAndCurativesRaoResultImpl.java:455]: and here too

**Review by phiedw — CHANGES_REQUESTED** [review]

File: data/rao-result/rao-result-api/src/test/java/com/farao_community/farao/data/rao_result_api/OptimizationStepsExecutedTest.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+    @Test
+    public void testFirstPrevFellback() {
+        OptimizationStepsExecuted optimizationStepsExecuted = OptimizationStepsExecuted.FIRST_PREVENTIVE_FELLBACK_TO_INITIAL_SITUATION;
+        assertFalse(optimizationStepsExecuted.hasRunSecondPreventive());
+        assertFalse(optimizationStepsExecuted.hasFallenBackToFirstPreventiveSituation());
+        assertTrue(optimizationStepsExecuted.hasFallenBackToInitialSituation());
+        assertFalse(optimizationStepsExecuted.isOverwritePossible(OptimizationStepsExecuted.SECOND_PREVENTIVE_IMPROVED_FIRST));
+        assertFalse(optimizationStepsExecuted.isOverwritePossible(OptimizationStepsExecuted.FIRST_PREVENTIVE_FELLBACK_TO_INITIAL_SITUATION));
+        assertFalse(optimizationStepsExecuted.isOverwritePossible(OptimizationStepsExecuted.SECOND_PREVENTIVE_FELLBACK_TO_FIRST_PREVENTIVE_SITUATION));
```
Comment (Godelaine) [inline data/rao-result/rao-result-api/src/test/java/com/farao_community/farao/data/rao_result_api/OptimizationStepsExecutedTest.java:40]: indentation (see other tests too)

File: data/rao-result/rao-result-impl/src/test/java/com/farao_community/farao/data/rao_result_impl/RaoResultImplTest.java [hunk 3]
```diff
@@ -205,4 +207,15 @@ public void testCostResults() {
 
         assertEquals(ComputationStatus.DEFAULT, raoResult.getComputationStatus());
     }
+
+    @Test
+    public void testRaoWentThroughSecondPrev() {
```
Comment (Godelaine) [inline data/rao-result/rao-result-impl/src/test/java/com/farao_community/farao/data/rao_result_impl/RaoResultImplTest.java:212]: i don't understand the name of the test : FIRST_PREV... doesn't mean that rao went through 2P, does it ?

File: data/rao-result/rao-result-json/src/main/java/com/farao_community/farao/data/rao_result_json/RaoResultJsonConstants.java [hunk 4]
```diff
@@ -108,6 +109,14 @@ private RaoResultJsonConstants() { }
     public static final String FALLBACK_STATUS = "fallback";
     public static final String FAILURE_STATUS = "failure";
 
+    // optimized steps executed by the RAO
+    public static final String OPTIMIZATION_STEPS_EXECUTED = "optimizationStepsExecuted";
+    public static final String FIRST_PREVENTIVE_ONLY = "The RAO went only through first preventive";
```
Comment (Godelaine) [inline data/rao-result/rao-result-json/src/main/java/com/farao_community/farao/data/rao_result_json/RaoResultJsonConstants.java:114]: only went through

**Review by Godelaine — APPROVED** [review]

**Pull request merged**